File: Frontend/src/app/services/auth.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable, tap } from 'rxjs';
import { LoginRequest } from '../interfaces/auth/request/login-request.model';
import { LoginResponse } from '../interfaces/auth/request/login-response.model';
import { RegisterRequest } from '../interfaces/auth/response/register-request.model';
import { RegisterResponse } from '../interfaces/auth/response/register-response.model';
import { CurrentUser } from '../interfaces/current user/current-user';


@Injectable({ providedIn: 'root' })
export class AuthService {
  private readonly apiUrl = 'http://localhost:8080/api/auth';
  private accessToken: string | null = null;
  private currentUser: CurrentUser | null = null;

  constructor(private http: HttpClient) {}

  login(request: LoginRequest): Observable<LoginResponse> {
    return this.http.post<LoginResponse>(`${this.apiUrl}/login`, request)
      .pipe(tap(response => this.setSession(response)));
  }

  register(request: RegisterRequest): Observable<RegisterResponse> {
    return this.http.post<RegisterResponse>(`${this.apiUrl}/register`, request);
  }

  refresh(): Observable<LoginResponse> {
    return this.http.post<LoginResponse>(`${this.apiUrl}/refresh`, {})
      .pipe(tap(response => this.setSession(response)));
  }

  logout(): void {
    this.http.post(`${this.apiUrl}/logout`, {}).subscribe();
    this.clearToken();
  }

  private setSession(response: LoginResponse): void {
  this.accessToken = response.accessToken;
  this.currentUser = {
    id: response.id,
    username: response.username,
    firstname: response.firstname,
    lastname: response.lastname,
    phone: response.phone,
    email: response.email,
    role: response.role
  };
}

  setToken(token: string): void {
    this.accessToken = token;
  }

  getToken(): string | null {
    return this.accessToken;
  }

  getCurrentUser(): CurrentUser | null {
    return this.currentUser;
  }

  clearToken(): void {
    this.accessToken = null;
    this.currentUser = null;
  }

  isAuthenticated(): boolean {
    return this.accessToken !== null;
  }
}

File: Backend/src/main/java/fr/eni/bookhub/auth/service/AuthService.java
package fr.eni.bookhub.auth.service;

import fr.eni.bookhub.auth.dto.request.LoginRequest;
import fr.eni.bookhub.auth.dto.response.LoginResponse;
import fr.eni.bookhub.auth.dto.request.RegisterRequest;
import fr.eni.bookhub.auth.dto.response.RegisterResponse;
import fr.eni.bookhub.exception.custom.ConflictException;
import fr.eni.bookhub.jwt.util.JwtUtil;
import fr.eni.bookhub.refreshToken.entity.RefreshToken;
import fr.eni.bookhub.refreshToken.service.RefreshTokenService;
import fr.eni.bookhub.user.dao.IUserDao;
import fr.eni.bookhub.user.entity.Role;
import fr.eni.bookhub.user.entity.User;
import lombok.AllArgsConstructor;
import org.springframework.security.authentication.AuthenticationManager;
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;

@AllArgsConstructor
@Service
public class AuthService {

    private final IUserDao userRepository;
    private final PasswordEncoder passwordEncoder;
    private final AuthenticationManager authenticationManager;
    private final JwtUtil jwtUtil;
    private final RefreshTokenService refreshTokenService;

    public RegisterResponse register(RegisterRequest request) {

        if (userRepository.findByEmail(request.getEmail()).isPresent()) {
            throw new ConflictException("Email already in use");
        }

        User user = User.builder()
                .firstname(request.getFirstname())
                .lastname(request.getLastname())
                .phone(request.getPhone())
                .username(request.getUsername())
                .email(request.getEmail())
                .password(passwordEncoder.encode(request.getPassword()))
                .role(Role.ROLE_USER)
                .build();

        User saved = userRepository.save(user);
        return RegisterResponse.fromUserEntity(saved);
    }

    public LoginResponse login(LoginRequest request) {

        authenticationManager.authenticate(
                new UsernamePasswordAuthenticationToken(
                        request.getEmail(),
                        request.getPassword()
                )
        );

        User user = userRepository.findByEmail(request.getEmail())
                .orElseThrow(() -> new RuntimeException("User not found"));

        String accessToken = jwtUtil.generateToken(user);
        String refreshToken = refreshTokenService.createToken(user);

        return LoginResponse.builder()
                .id(user.getId())
                .firstname(user.getFirstname())
                .lastname(user.getLastname())
                .username(user.getDisplayUsername())
                .phone(user.getPhone())
                .email(user.getEmail())
                .role(user.getRole())
                .accessToken(accessToken)
                .refreshToken(refreshToken)
                .build();
    }

    public LoginResponse refresh(String rawRefreshToken) {

        RefreshToken oldToken = refreshTokenService.validateToken(rawRefreshToken);
        User user = oldToken.getUser();

        String newAccessToken = jwtUtil.generateToken(user);
        String newRefreshToken = refreshTokenService.rotateToken(oldToken);

        return LoginResponse.builder()
                .id(user.getId())
                .email(user.getEmail())
                .firstname(user.getFirstname())
                .lastname(user.getLastname())
                .username(user.getDisplayUsername())
                .phone(user.getPhone())
                .role(user.getRole())
                .accessToken(newAccessToken)
                .refreshToken(newRefreshToken)
                .build();
    }

    public void logout(String rawRefreshToken) {
        if (rawRefreshToken != null) {
            refreshTokenService.revokeToken(rawRefreshToken);
        }
    }
}


